Building DataMySQL fails

Poco 1.12.1
Ubuntu 20.04.4 LTS
CMake 3.16.3

Building DataMySQL fails

Scanning dependencies of target DataMySQL

[ 94%] Building CXX object Data/MySQL/CMakeFiles/DataMySQL.dir/src/Binder.cpp.o

[ 94%] Building CXX object Data/MySQL/CMakeFiles/DataMySQL.dir/src/Connector.cpp.o

[ 94%] Building CXX object Data/MySQL/CMakeFiles/DataMySQL.dir/src/Extractor.cpp.o

/home/poco/poco-1.12.1-all/Data/MySQL/src/Extractor.cpp: In member function ‘bool Poco::Data::MySQL::Extractor::extractJSON(std::size_t)’:

/home/poco/poco-1.12.1-all/Data/MySQL/src/Extractor.cpp:303:21: error: ‘MYSQL_TYPE_JSON’ was not declared in this scope; did you mean ‘MYSQL_TYPE_BLOB’?
  303 |  row->buffer_type = MYSQL_TYPE_JSON;
      |                     ^~~~~~~~~~~~~~~
      |                     MYSQL_TYPE_BLOB

make[2]: *** [Data/MySQL/CMakeFiles/DataMySQL.dir/build.make:89: 

Data/MySQL/CMakeFiles/DataMySQL.dir/src/Extractor.cpp.o] Error 1

make[1]: *** [CMakeFiles/Makefile2:5135: Data/MySQL/CMakeFiles/DataMySQL.dir/all] Error 2

make: *** [Makefile:163: all] Error 2

/Users/andy/project/CLion/poco-1.12.1/Data/MySQL/src/Extractor.cpp:303:21: error: use of undeclared identifier 'MYSQL_TYPE_JSON'
        row->buffer_type = MYSQL_TYPE_JSON;
                           ^
1 error generated.
make[2]: *** [Data/MySQL/CMakeFiles/DataMySQL.dir/src/Extractor.cpp.o] Error 1
make[1]: *** [Data/MySQL/CMakeFiles/DataMySQL.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

[ 97%] Building CXX object Net/testsuite/CMakeFiles/Net-testrunner.dir/src/MailMessageTest.cpp.o
duplicate symbol 'Poco::Crypto::Envelope::getContent() const' in:
    CMakeFiles/Crypto-testrunner.dir/src/EnvelopeTest.cpp.o
    ../../lib/libPocoCryptod.a(Envelope.cpp.o)
duplicate symbol 'Poco::Crypto::Envelope::setContent(std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > const&)' in:
    CMakeFiles/Crypto-testrunner.dir/src/EnvelopeTest.cpp.o
    ../../lib/libPocoCryptod.a(Envelope.cpp.o)
ld: 2 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/Crypto-testrunner] Error 1
make[1]: *** [Crypto/testsuite/CMakeFiles/Crypto-testrunner.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

--- a/Data/MySQL/include/Poco/Data/MySQL/Extractor.h
+++ b/Data/MySQL/include/Poco/Data/MySQL/Extractor.h
@@ -324,9 +324,9 @@ private:
 	bool realExtractFixed(std::size_t pos, enum_field_types type, void* buffer, bool isUnsigned = false);
 
 	bool extractLongLOB(std::size_t pos);
-
+#ifdef POCO_MYSQL_JSON
 	bool extractJSON(std::size_t pos);
-
+#endif
 	// Prevent VC8 warning "operator= could not be generated"
 	Extractor& operator=(const Extractor&);
 

--- a/Data/MySQL/include/Poco/Data/MySQL/MySQL.h
+++ b/Data/MySQL/include/Poco/Data/MySQL/MySQL.h
@@ -62,5 +62,17 @@
 	#endif
 #endif
 
+//
+// Detect support for JSON data type
+//
+#if defined(MARIADB_VERSION_ID)
+	#if MARIADB_VERSION_ID > 100207
+		#define POCO_MYSQL_JSON
+	#endif
+#else
+	#if MYSQL_VERSION_ID > 50708
+		#define POCO_MYSQL_JSON
+	#endif
+#endif
 
 #endif // MySQL_MySQL_INCLUDED

--- a/Data/MySQL/include/Poco/Data/MySQL/MySQL.h
+++ b/Data/MySQL/include/Poco/Data/MySQL/MySQL.h
@@ -66,11 +66,11 @@
 // Detect support for JSON data type
 //
 #if defined(MARIADB_VERSION_ID)
-	#if MARIADB_VERSION_ID > 100207
+	#if MARIADB_VERSION_ID >= 100207
 		#define POCO_MYSQL_JSON
 	#endif
 #else
-	#if MYSQL_VERSION_ID > 50708
+	#if MYSQL_VERSION_ID >= 50708
 		#define POCO_MYSQL_JSON
 	#endif
 #endif

--- a/Data/MySQL/src/Extractor.cpp
+++ b/Data/MySQL/src/Extractor.cpp
@@ -128,12 +128,16 @@ bool Extractor::extract(std::size_t pos, std::string& val)
 
 	//mysql reports TEXT types as FDT_BLOB when being extracted
 	MetaColumn::ColumnDataType columnType = _metadata.metaColumn(static_cast<Poco::UInt32>(pos)).type();
+#ifdef POCO_MYSQL_JSON
 	if (columnType != Poco::Data::MetaColumn::FDT_STRING && columnType != Poco::Data::MetaColumn::FDT_BLOB && columnType != Poco::Data::MetaColumn::FDT_JSON)
+#else
+	if (columnType != Poco::Data::MetaColumn::FDT_STRING && columnType != Poco::Data::MetaColumn::FDT_BLOB)
+#endif
 		throw MySQLException("Extractor: not a string");
-
+#ifdef POCO_MYSQL_JSON
 	if (columnType == Poco::Data::MetaColumn::FDT_JSON && !extractJSON(pos))
 		return false;
-
+#endif
 	if (columnType == Poco::Data::MetaColumn::FDT_BLOB && !extractLongLOB(pos))
 		return false;
 
@@ -291,6 +295,7 @@ bool Extractor::extractLongLOB(std::size_t pos)
 	return true;
 }
 
+#ifdef POCO_MYSQL_JSON
 bool Extractor::extractJSON(std::size_t pos)
 {
 	// JSON columns are fetched with a zero-length
@@ -306,6 +311,7 @@ bool Extractor::extractJSON(std::size_t pos)
 
 	return true;
 }
+#endif
 
 //////////////
 // Not implemented

--- a/Data/MySQL/src/ResultMetadata.cpp
+++ b/Data/MySQL/src/ResultMetadata.cpp
@@ -72,7 +72,9 @@ namespace
 		case MYSQL_TYPE_MEDIUM_BLOB:
 		case MYSQL_TYPE_LONG_BLOB:
 		case MYSQL_TYPE_BLOB:
+#ifdef POCO_MYSQL_JSON
 		case MYSQL_TYPE_JSON:
+#endif		
 			return field.length;
 
 		default:

--- a/Data/MySQL/testsuite/src/MySQLTest.cpp
+++ b/Data/MySQL/testsuite/src/MySQLTest.cpp
@@ -487,6 +487,7 @@ void MySQLTest::testLongTEXT()
 	_pExecutor->longText();
 }
 
+#ifdef POCO_MYSQL_JSON
 void MySQLTest::testJSON()
 {
 	if (!_pSession) fail("Test not available.");
@@ -494,7 +495,7 @@ void MySQLTest::testJSON()
 	recreatePersonJSONTable();
 	_pExecutor->json();
 }
-
+#endif
 
 void MySQLTest::testUnsignedInts()
 {
@@ -796,7 +797,7 @@ void MySQLTest::recreatePersonLongBLOBTable()
 	catch(StatementException& se){ std::cout << se.displayText() << std::endl; fail ("recreatePersonLongBLOBTable()"); }
 }
 
-
+#ifdef POCO_MYSQL_JSON
 void MySQLTest::recreatePersonJSONTable()
 {
 	dropTable("Person");
@@ -804,7 +805,7 @@ void MySQLTest::recreatePersonJSONTable()
 	catch (ConnectionException& ce) { std::cout << ce.displayText() << std::endl; fail("recreatePersonJSONTable()"); }
 	catch (StatementException& se) { std::cout << se.displayText() << std::endl; fail("recreatePersonJSONTable()"); }
 }
-
+#endif
 
 void MySQLTest::recreateIntsTable()
 {

--- a/Data/MySQL/testsuite/src/MySQLTest.h
+++ b/Data/MySQL/testsuite/src/MySQLTest.h
@@ -81,7 +81,9 @@ public:
 	void testBLOBStmt();
 	void testLongBLOB();
 	void testLongTEXT();
+#ifdef POCO_MYSQL_JSON
 	void testJSON();
+#endif
 
 	void testUnsignedInts();
 	void testFloat();
@@ -122,7 +124,9 @@ private:
 	void recreatePersonTimeTable();
 	void recreatePersonTimestampTable();
 	void recreatePersonLongBLOBTable();
+#ifdef POCO_MYSQL_JSON
 	void recreatePersonJSONTable();
+#endif
 	void recreateStringsTable();
 	void recreateIntsTable();
 	void recreateUnsignedIntsTable();

--- a/Data/MySQL/testsuite/src/SQLExecutor.cpp
+++ b/Data/MySQL/testsuite/src/SQLExecutor.cpp
@@ -1513,6 +1513,7 @@ void SQLExecutor::longText()
 	poco_assert (longTextRes == biography);
 }
 
+#ifdef POCO_MYSQL_JSON
 void SQLExecutor::json()
 {
 	std::string funct = "json()";
@@ -1537,7 +1538,7 @@ void SQLExecutor::json()
 	catch (StatementException& se) { std::cout << se.displayText() << std::endl; fail(funct); }
 	poco_assert(res == biography);
 }
-
+#endif
 
 void SQLExecutor::tuples()
 {

--- a/Data/MySQL/testsuite/src/SQLExecutor.h
+++ b/Data/MySQL/testsuite/src/SQLExecutor.h
@@ -86,7 +86,9 @@ public:
 	void timestamp();
 	void longBlob();
 	void longText();
+#ifdef POCO_MYSQL_JSON
 	void json();
+#endif	
 	void unsignedInts();
 	void floats();
 	void doubles();

